### PR TITLE
Feature/docker hardening

### PR DESCRIPTION
**Pull request by jason-fox**
This PR fixes several issues raised by [docker/docker-bench-security](https://github.com/docker/docker-bench-security)

-   Add missing `HEALTHCHECK` - currently just a ping to `/version`, but could be expanded
-   Add standard metadata build `LABEL` (see [opencontainers.org](https://github.com/opencontainers/opencontainers.org))
-   Remove **deprecated** `MAINTAINER` - replace with `LABEL` 
-   Create two-stage build process  using a ~**Distroless** build~ **Alpine** build:
     -   ~Remove [disto](https://medium.com/@dwdraju/distroless-is-for-security-if-not-for-size-6eac789f695f)~ Use **Alpine**
     -    Run as anonymous user 
     -    Set `NODE_ENV` to development/production
-   Ensure that Dockerfile continues to follow best practice by linting on CI. Using [Hadolint](https://github.com/hadolint/hadolint)
-   Remove Docker Secret support from `entrypoint.sh` since it should be processed by Node not Bash. (see https://github.com/telefonicaid/iotagent-node-lib/pull/885)
-   Remove the `entrypoint.sh`, the database check and seeding is now within the Keyrock code itself.
-   Remove the `postfix` dependency, this should be in a separate docker container as it violates the _one-concern-per-container_ principle and amend documentation to show how to set up a proxy mailer instead

This reduces the image size from 556MB to 208MB.

#### Rationale for Alpine

It will be difficult to go full **Distroless** in one PR because of the various dependencies within the Keyrock docker container, so I've opted to do the following:

- Leave the sequelize database creation as `npm` statements  (requires `bash` and `npm`)

### CODE REVIEW

**Comment by jason-fox** [comment]
This leaves an issue regarding seeding the database.

1) Distroless Build - lightweight, secure, no bash
```console
 docker build -t fiware/idm .  
```

2) Debian Build - just builds the resource files in a Node container.
```console
docker build -t fiware/idm --target=seeder .  
```

This includes the standard entrypoint (at the cost of quadrupling the Docker image size) 😄

**Comment by jason-fox** [comment]
It would be better to allow an exec command to do the seeding.

```console
docker exec fiware-keyrock /nodejs/bin/node ./bin/healthcheck
```

Successfully runs the HEALTHCHECK

```console
docker exec fiware-keyrock /nodejs/bin/node ./bin/database
```

Attempts to run `sequelize`, but currently fails - I'm not familiar with Sequelize so I'd need you to fix this yourself.

**Comment by jason-fox** [comment]
An alternative would be to provide the SQL statements `backup.sql`  for the seeded database as part of the MySQL docker

see: https://stackoverflow.com/questions/25920029/setting-up-mysql-and-importing-dump-within-dockerfile

```yaml
  mysql-db:
    restart: always
    image: mysql:${MYSQL_DB_VERSION}
    hostname: mysql-db
    container_name: db-mysql
    expose:
      - "${MYSQL_DB_PORT}"
    ports:
      - "${MYSQL_DB_PORT}:${MYSQL_DB_PORT}" 
    environment:
      - "MYSQL_ROOT_PASSWORD_FILE=/run/secrets/my_secret_data"
      - "MYSQL_ROOT_HOST=172.18.1.5" # Allow Keyrock to access this database
    volumes:
      - mysql-db:/var/lib/mysql
      - ./mysql-data:/docker-entrypoint-initdb.d/:ro
```

File: bin/database.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  })
+  .catch(err => {
+    // added '3D000' error code for postgres cause not referenced in dialect as 'ER_BAD_DB_ERROR'
+    if (
+      err.original.code === 'ER_BAD_DB_ERROR' ||
+      err.original.code === '3D000'
+    ) {
+      exec('sequelize --env database db:create', function(error) {
+        if (error) {
+          debug('Unable to create database: ', err);
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply bin/database.js]: This makes sense as the *Production* build is Distroless, and therefore won't contain `/bin/bash` (or 500MB of other tools.)

File: bin/database.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+          process.exit();
+        }
+
+        debug('Database created');
+
+        exec('sequelize --env database db:migrate', function(error) {
+          if (error) {
+            debug('Unable to migrate database: ', err);
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply bin/database.js]: Fixed: 3bf60c8

**Comment by jason-fox** [comment]
Currently the main  `docker build -t fiware/idm .` build image won't offer a working database seeder. To seed a database you can do the following:

Build with bash:

```console
 docker build -t fiware/idm . --target=builder
```

Create DB

```
docker exec fiware-keyrock node ./bin/database -e IDM_DB_HOST=mysql-db -e IDM_DB_PASS=secret
```


Seeding also works with the `npm run-scripts` from the source code.

**Comment by jason-fox** [comment]
I've finally got the unit tests running and I've removed several redundant `cloned config.js` files. I think this PR now ready for review.

**Comment by apozohue10** [comment]
Hi Jason,

I have reviewed your PR and I think it is great, but I have some considerations before accepting it.

I see that you have removed the config.js.template. It is true that in a Docker deployment it does not make sense as it takes up space. However, It is quite useful for development in the source code. 

I also see that the create_db, migrate_db and seed_db functions have been also automatized when running Keyrock. Again, I think it makes sense for the docker deployment but not for the source code. As a developer, I would prefer to manually control the creation, migration and seed of the DB rather than an automatized execution of these functions.

If you edit these two issues I will accept it ASAP.

Kind regards and thank you

**Comment by jason-fox** [comment]
@apozohue10 

> I see that you have removed the `config.js.template`. It is true that in a Docker deployment it does not make sense as it takes up space. 

`config.js.template` is now just `config.js`. This is holding all the default values.

> However, It is quite useful for development in the source code.

I'm not sure how do you want me to proceed here? Just add another copy of `config.js`  back in or what?

**Comment by jason-fox** [comment]
> I also see that the create_db, migrate_db and seed_db functions have been also automatized when running Keyrock. Again, I think it makes sense for the docker deployment but not for the source code. As a developer, I would prefer to manually control the creation, migration and seed of the DB rather than an automatized execution of these functions.

The `npm run create_db| migrate_db|seed_db` still exist, so it is possible to run them directly. The code _as-is_ is just running the seeder when no database is found.

To give more control I guess the right way is to set an `ENV` in the `Dockerfile` and default  to something like `IDM_SEED_DB=true` . The `ENV` wouldn't exist in a local installation and therefore wouldn't run automatically.

**Comment by jason-fox** [comment]
> To give more control I guess the right way is to set an `ENV` in the `Dockerfile` and default to something like `IDM_SEED_DB=true` . The `ENV` wouldn't exist in a local installation and therefore wouldn't run automatically.
> 
> Basically add a flag around _Attempt to seed_ 

This could also mean that a subsequent PR could re-add in *Distroless* with a default `IDM_DB_SEED=false` (since `npm` won't be supported on Distroless). You could even have separate builds for node-slim, alpine and distroless. Since such a change would be significant, I don't want to introduce it to this PR to keep this review manageable.

**Comment by apozohue10** [comment]
Hi Jason, 

that's it. We think is better to conserve the copy of config.js for development.

Regarding the IDM_DB_SEED I think is perfect!

**Comment by jason-fox** [comment]
> it is better to conserve the copy of `config.js` for development.

Fixed: 32660f4

**Pull request merged**